extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/SegmentLoadWaiter.java

```diff
@@ -0,0 +1,392 @@
+
+    String response = brokerClient.sendQuery(request);
+
+    if (response.trim().isEmpty()) {
+      // If no segments are returned for a version, all segments have been dropped by a drop rule.
+      return new VersionLoadStatus(0, 0, 0, 0, 0);
+    } else {
+      return objectMapper.readValue(response, VersionLoadStatus.class);
+    }
+  }
+
+  /**
+   * Returns the current status of the load.
+   */
+  public SegmentLoadWaiterStatus status()
+  {
+    return status.get();
+  }
+
+  public static class SegmentLoadWaiterStatus
+  {
+    private final State state;
+    private final DateTime startTime;
+    private final long duration;
+    private final int totalSegments;
+    private final int usedSegments;
+    private final int precachedSegments;
+    private final int asyncOnlySegments;
+    private final int pendingSegments;
+    private final int unknownSegments;
+
+    @JsonCreator
+    public SegmentLoadWaiterStatus(
+        @JsonProperty("state") SegmentLoadWaiter.State state,
+        @JsonProperty("startTime") @Nullable DateTime startTime,
+        @JsonProperty("duration") long duration,
+        @JsonProperty("totalSegments") int totalSegments,
+        @JsonProperty("usedSegments") int usedSegments,
+        @JsonProperty("precachedSegments") int precachedSegments,
+        @JsonProperty("asyncOnlySegments") int asyncOnlySegments,
+        @JsonProperty("pendingSegments") int pendingSegments,
+        @JsonProperty("unknownSegments") int unknownSegments
+    )
+    {
+      this.state = state;
+      this.startTime = startTime;
+      this.duration = duration;
+      this.totalSegments = totalSegments;
+      this.usedSegments = usedSegments;
+      this.precachedSegments = precachedSegments;
+      this.asyncOnlySegments = asyncOnlySegments;
+      this.pendingSegments = pendingSegments;
+      this.unknownSegments = unknownSegments;
+    }
+
+    @JsonProperty
+    public SegmentLoadWaiter.State getState()
+    {
+      return state;
+    }
+
+    @Nullable
+    @JsonProperty
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    public DateTime getStartTime()
+    {
+      return startTime;
+    }
+
+    @JsonProperty
+    public long getDuration()
+    {
+      return duration;
+    }
+
+    @JsonProperty
+    public long getTotalSegments()
+    {
+      return totalSegments;
+    }
+
+    @JsonProperty
+    public int getUsedSegments()
+    {
+      return usedSegments;
+    }
+
+    @JsonProperty
+    public int getPrecachedSegments()
+    {
+      return precachedSegments;
+    }
+
+    @JsonProperty
+    public int getAsyncOnlySegments()
+    {
+      return asyncOnlySegments;
+    }
+
+    @JsonProperty
+    public int getPendingSegments()
+    {
+      return pendingSegments;
+    }
+
+    @JsonProperty
+    public int getUnknownSegments()
+    {
+      return unknownSegments;
+    }
+  }
+
+  public enum State
+  {
+    INIT,
```

Comment:
Could you add more dev notes to this enum.

Reply:
Added